pkg/kubelet/kuberuntime/helpers_linux.go

```diff
@@ -65,3 +69,20 @@ func milliCPUToQuota(milliCPU int64, period int64) (quota int64) {
 
 	return
 }
+
+// sharesToMilliCPU converts CpuShares (cpu.shares) to milli-CPU value
+func sharesToMilliCPU(shares int64) int64 {
```

Comment:
We bottomline it to 2 when configuring it. 
https://github.com/kubernetes/kubernetes/pull/102884/files#diff-da1da20ffb0529c88d866cbb23e93f6f30a89b5906208ce0daff5b667b1ebc5eR44

Reply:
Sorry, can you link to the right place? :)

Reply:
my bad. https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/kubelet/kuberuntime/helpers_linux.go#L40

We never configure less than minShares (2) so this is a reasonable inverse.

Reply:
The problem is that `1mCPU` maps to 2 shares, but when mapping back 2 shares maps to `2mCPU`.

Will this cause problem? Because in theory it is possible that desired state and actual state will never match.

Reply:
Yes, this is a problem. Even if we configured 1m/1m requests/limits, the actual configuration shows 2m/10m.
This change seems to have a role: https://github.com/kubernetes/kubernetes/pull/23143

This will result in SyncPod repeatedly calling CRI to "correct" the problem. I'm not sure what a good fix would be. Should we min cpu.requests/limits to 2m/10m in spec? Or hack it in SyncPod (can cause user confusion)

```
 containerStatuses:
  - containerID: docker://45782ba2bb77549be2b1ff7b1b7a52116f048f2186bf6b084c732e53c0942e99
    image: busybox:latest
    imageID: docker-pullable://busybox@sha256:930490f97e5b921535c153e0e7110d251134cc4b72bbb8133c6a5065cc68580d
    lastState: {}
    name: stress
    ready: true
    resources:
      limits:
        cpu: 10m
        memory: 500Mi
      requests:
        cpu: 2m
        memory: 500Mi
    resourcesAllocated:
      cpu: 1m
      memory: 500Mi
```

Reply:
Yeah, that is a similar issue with CPU quota.

> Should we min cpu.requests/limits to 2m/10m in spec? 

That seems to be a breaking change.

Maybe we can detect that: if both desired and actual quota/share is the min value, we keep the desired mCPU.
* If desired > min value: it doesn't matter what the resource allocated is, the underlying quota/share is min value anyway;
* If desired = min value: it won't cause confusion, and the underlying quota/share is min value anyway.

That is one potential solution, but I feel like this needs wider discussion to collect feedback

Reply:
So handle this in API? CC: @thockin .. I think we need Tim's input on this.

Imho, failing this in validation is the right way to do this since system does not allow requests/limits <= 2m/10m. But that would break existing yamls. The "pseudo limit-ranger" is the next best option to address this hack. Hiding this in kubelet is the last option.

Reply:
We can't break users.  If it is valid to specif `1m` for CPU today (even if internally we make that the same as `2m`) then we MUST continue to allow that in API.

Options:
a) change the logic to make `1m` mean `1m`.  I don't have all the history there.
b) make it the runtime's problem to preserve and return the user-provided value, even if it mutates that when configuring the OS
c) provide a way for the kubelet to to ask the CRI for lower-bounds and not try to reconcile beyond that.

If I understand correctly, I think option (b) is right-est.  The choice of lower-bound is an implementation detail of CRI.

Reply:
@vinaykul Is this addressed?

Reply:
No. I don't know what a safe fix is here, I have not deep-dived and fully digested the reasons for the 2m/10m hack. 

I'm tracking this as an issue to fix before beta (See https://github.com/kubernetes/enhancements/issues/1287#issue-503823628) so I have the time to talk to folks and really understand the reasons why this was done.

Reply:
In that case, please:
1) Add a TODO, and make sure this is addressed in beta;
2) Document a known issue for the alpha feature.

Reply:
Sure. TODO coming in next commit. 
```diff
diff --git a/pkg/kubelet/kubelet.go b/pkg/kubelet/kubelet.go
index ec810cedd84..cac6cbf9e14 100644
--- a/pkg/kubelet/kubelet.go
+++ b/pkg/kubelet/kubelet.go
@@ -1724,6 +1724,8 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
        if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
                // Handle pod resize here instead of doing it in HandlePodUpdates because
                // this conveniently retries any Deferred resize requests
+               // TODO(vinaykul): Investigate doing this in HandlePodUpdates + periodic SyncLoop scan
+               //     See: https://github.com/kubernetes/kubernetes/pull/102884#discussion_r663160060
                if !kl.podIsTerminated(pod) && !kubepod.IsStaticPod(pod) {
                        kl.handlePodResourcesResize(pod)
                }
diff --git a/pkg/kubelet/kuberuntime/helpers_linux.go b/pkg/kubelet/kuberuntime/helpers_linux.go
index 00b133a7dd0..882098e395a 100644
--- a/pkg/kubelet/kuberuntime/helpers_linux.go
+++ b/pkg/kubelet/kuberuntime/helpers_linux.go
@@ -71,6 +71,8 @@ func milliCPUToQuota(milliCPU int64, period int64) (quota int64) {
 }
 
 // sharesToMilliCPU converts CpuShares (cpu.shares) to milli-CPU value
+//TODO(vinaykul): Address issue that sets min req/limi to 2m/10m before beta
+//  See: https://github.com/kubernetes/kubernetes/pull/102884#discussion_r662552642
 func sharesToMilliCPU(shares int64) int64 {
        milliCPU := int64(0)
        if shares >= minShares {
root@fw0000359:~/go/src/k8s.io/kubernetes-rfpvs/pkg/kubelet#
```
I'm tracking this issue on KEP thread as well https://github.com/kubernetes/enhancements/issues/1287#issue-503823628

Reply:
for these TODO, it would be good if they all have a reference to associated FeatureGate name so its easy to grep and find during beta phase.

Reply:
TODOs updated with featuregate.

Reply:
Apologies for the slight tangent but if I have understood correctly that `2m` is treated as the effective minimum CPU request today regardless of these changes, would anyone be able to point me at the history / a reference for that? Documentation, discussion or code is fine. (We use `1m` requests quite often.)

Reply:
Sorry for the delayed response. I have not had the time to track down the history on this. Perhaps @vishh can give us some context on why we have these minimum values in the first place.

Reply:
@jbg see the following:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/helpers_linux.go#L41
https://github.com/torvalds/linux/blob/0bddd227f3dc55975e2b8dfa7fc6f959b062a2c7/kernel/sched/sched.h#L427-L428

Reply:
MinShares=2 comes from kernel clamping the min value to 2 in order to avoid arithmetic problems per [kernel code comments](https://github.com/torvalds/linux/blob/0bddd227f3dc55975e2b8dfa7fc6f959b062a2c7/kernel/sched/sched.h#L419-L428).

This deserves a separate beta-blocker issue with action item for tracking so here it is: https://github.com/kubernetes/kubernetes/issues/114123 

We cannot break existing yaml's that may request 1m CPU. My suggestion is to handle this in kubelet by setting status.resources.requests.cpu = status.resourcesAllocated when spec.resources.request.cpu <= 2m and  kubeContainerStatus.cpu.shares = 2 (this is the value of cpu.shares reported by runtime for both 1m and 2m CPU requests due to our clamp).

Any other suggestions are welcome.